code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/process/LoadBalancerServiceInstanceStopPreListener.java

```diff
@@ -0,0 +1,149 @@
+import io.cattle.platform.core.constants.InstanceConstants;
+import io.cattle.platform.core.constants.ServiceConstants;
+import io.cattle.platform.core.dao.GenericMapDao;
+import io.cattle.platform.core.dao.InstanceDao;
+import io.cattle.platform.core.model.Instance;
+import io.cattle.platform.core.model.InstanceHostMap;
+import io.cattle.platform.core.model.Service;
+import io.cattle.platform.core.model.ServiceExposeMap;
+import io.cattle.platform.engine.handler.HandlerResult;
+import io.cattle.platform.engine.handler.ProcessPreListener;
+import io.cattle.platform.engine.process.ProcessInstance;
+import io.cattle.platform.engine.process.ProcessState;
+import io.cattle.platform.eventing.model.EventVO;
+import io.cattle.platform.json.JsonMapper;
+import io.cattle.platform.object.util.DataAccessor;
+import io.cattle.platform.object.util.ObjectUtils;
+import io.cattle.platform.process.common.handler.AgentBasedProcessHandler;
+import io.cattle.platform.servicediscovery.service.ServiceDiscoveryService;
+import io.cattle.platform.util.type.CollectionUtils;
+import io.cattle.platform.util.type.Priority;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import javax.inject.Inject;
+
+import org.apache.commons.lang3.StringUtils;
+
+
+public class LoadBalancerServiceInstanceStopPreListener extends AgentBasedProcessHandler implements ProcessPreListener, Priority {
+    
+    @Inject
+    JsonMapper jsonMapper;
+    
+    @Inject
+    ServiceDiscoveryService sdService;
+    
+    @Inject
+    InstanceDao instanceDao;
+    
+    @Inject
+    GenericMapDao mapDao;
+
+    @Override
+    public HandlerResult handle(ProcessState state, ProcessInstance process) {
+        Instance instance = (Instance) state.getResource();
+        if (instance == null) {
+            return null;
+        }
+        //container.stop -> find container’s service -> find all the LB services of the account of the container service -> filter out those services having lbConfig.portRules having serviceId = container’s service
+        ServiceExposeMap smap = objectManager.findAny(ServiceExposeMap.class,
+                    SERVICE_EXPOSE_MAP.INSTANCE_ID, instance.getId(),
+                    SERVICE_EXPOSE_MAP.REMOVED, null);
+        if (smap == null) {
+            return null;
+        }
+        Service service = getObjectManager().loadResource(Service.class, smap.getServiceId());
+        if (service == null) {
+            return null;
+        }
+        List<Service> activeLbServices = lookupAllActiveLBServices(service);
+        List<? extends Instance> lbInstances = getLBServiceInstances(activeLbServices);
+        //send drain event to each lbInstance and wait for reply upto timeout.
+
+        for (Instance lbInstance : lbInstances) {
+            for (InstanceHostMap ihm : mapDao.findNonPurged(InstanceHostMap.class, Instance.class, lbInstance.getId())) {
+                Object agentResource = getObjectByRelationship("host", ihm);
```

Comment:
Dont do this line. Actually, you dont need this inner ihm loop at all. I think you just want to pass lbInstance to lookupAgent